### DISCUSSION TITLE
Considering alternative zlib implementations?

### DISCUSSION THREAD

**Discussion by shoffmeister** [rui314/mold#602] [Ideas]
A recent commit vendored zlib. It might be worthwhile considering replacing it with the Cloudflare fork (https://github.com/cloudflare/zlib) to (most likely) attain superior performance properties.

See https://aws.amazon.com/blogs/opensource/improving-zlib-cloudflare-and-comparing-performance-with-other-zlib-forks/ for benchmarking of an AWS version that was later upstreamed into the Cloudflare fork. I believe that this fork also contains adjustments from a Chromium vendored fork.

As far as I can tell from public sources (and limited personal experience), the performance claims of this fork hold.

Note that there is also https://github.com/zlib-ng/zlib-ng which claims to be more modern, incorporate all the Cloudflare performance, and which seems to be under reasonably active development. I have no experience with that fork.

**Comment by rui314** [comment]
Thank you for the suggestion. Looks like the Cloudflare's fork indeed seems a bit better than the official one, but I don't think we want to replace it at the moment. It's because we will use zlib compression on Windows only when we will be doing cross-compilation for ELF with compressed debug info. Linking Linux binaries on Windows isn't a very likely scenario, and with compressed debug info should be even rarer. I added zlib for the sake of completeness so that mold behaves the same way both on Linux and Windows, but I don't think it'll be used heavily.
